components/apimgt/org.wso2.carbon.apimgt.persistence.mongodb/src/main/java/org/wso2/carbon/apimgt/persistence/mongodb/MongoDBPersistenceImpl.java

```diff
@@ -0,0 +1,813 @@
+import org.wso2.carbon.apimgt.persistence.exceptions.ThumbnailPersistenceException;
+import org.wso2.carbon.apimgt.persistence.exceptions.WSDLPersistenceException;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.mongodb.client.model.Aggregates.group;
+import static com.mongodb.client.model.Aggregates.match;
+import static com.mongodb.client.model.Aggregates.project;
+import static com.mongodb.client.model.Aggregates.unwind;
+import static com.mongodb.client.model.Filters.and;
+import static com.mongodb.client.model.Filters.eq;
+import static com.mongodb.client.model.Projections.include;
+import static com.mongodb.client.model.Updates.combine;
+import static com.mongodb.client.model.Updates.pull;
+import static com.mongodb.client.model.Updates.push;
+import static com.mongodb.client.model.Updates.set;
+import static com.mongodb.client.model.Projections.exclude;
+
+public class MongoDBPersistenceImpl implements APIPersistence {
+
+    private static final Log log = LogFactory.getLog(MongoDBPersistenceImpl.class);
+    private static Map<String, Boolean> indexCheckCache = new HashMap<>();
+
+    @Override
+    public PublisherAPI addAPI(Organization org, PublisherAPI publisherAPI) throws APIPersistenceException {
+
+        MongoCollection<MongoDBPublisherAPI> collection = getPublisherCollection(org.getName());
+        publisherAPI.setCreatedTime(String.valueOf(new Date().getTime()));
+        MongoDBPublisherAPI mongoDBPublisherAPI = MongoAPIMapper.INSTANCE.toMongoDBPublisherApi(publisherAPI);
+        InsertOneResult insertOneResult = collection.insertOne(mongoDBPublisherAPI);
+        MongoDBPublisherAPI createdDoc = collection.find(eq("_id", insertOneResult.getInsertedId())).first();
+
+        //Create atlas search index in async manner
+        Runnable runnable = () -> {
+            isIndexCreated(org.getName());
+        };
+        Thread thread = new Thread(runnable);
+        thread.start();
+        return MongoAPIMapper.INSTANCE.toPublisherApi(createdDoc);
+    }
+
+    private Boolean isIndexCreated(String organizationName) {
+        //Check if index is created flag exists in cache
+        Boolean indexCacheCheck = indexCheckCache.get(organizationName);
+        if (indexCacheCheck != null && indexCacheCheck) {
+            return true;
+        }
+        //Call mongodb atlas rest api to check if index exists
+        boolean isIndexEmpty = MongoDBAtlasAPIConnector.getInstance().getIndexes(organizationName).isEmpty();
+        if (!isIndexEmpty) {
+            indexCheckCache.put(organizationName, true);
+            return true;
+        }
+
+        // Should we retry if we failed?
```

Comment:
we can retry a predefined count (3) between a certain delay (5s) may be? ideally we should get those from a the configuration.